src/main/kotlin/apply/application/EvaluationService.kt

```diff
@@ -0,0 +1,86 @@
+package apply.application
+
+import apply.domain.evaluation.Evaluation
+import apply.domain.evaluation.EvaluationRepository
+import apply.domain.evaluation.dto.EvaluationResponse
+import org.springframework.data.repository.findByIdOrNull
+import org.springframework.stereotype.Service
+import javax.annotation.PostConstruct
+import javax.transaction.Transactional
+
+@Transactional
+@Service
+class EvaluationService(
+    private val evaluationRepository: EvaluationRepository,
+    private val recruitmentService: RecruitmentService
+) {
+    private fun findAll(): List<Evaluation> {
+        return evaluationRepository.findAll()
+    }
+
+    private fun findById(id: Long): Evaluation? {
+        if (id == 0L) return null
+
+        return evaluationRepository.findByIdOrNull(id) ?: throw IllegalArgumentException("해당 id의 평가를 찾을 수 없습니다.")
+    }
+
+    fun findAllWithRecruitment(): List<EvaluationResponse> {
+        return findAll().map {
+            EvaluationResponse(
+                it.id,
+                it.title,
+                it.description,
+                recruitmentService.getById(it.recruitment).title,
+                it.recruitment,
+                findById(it.beforeEvaluation)?.title ?: "이전 평가 없음",
+                it.beforeEvaluation
+            )
+        }.toList()
+    }
+
+    fun deleteById(id: Long) {
+        evaluationRepository.deleteById(id)
+
+        resetBeforeEvaluationContain(id)
+    }
+
+    private fun resetBeforeEvaluationContain(id: Long) {
+        evaluationRepository.findAll()
+            .firstOrNull { it.hasSameBeforeEvaluationWith(id) }?.resetBeforeEvaluation()
```

Comment:
이전 평가와 연결된 모든 평가를 수정해야 하지 않을까요?

Reply:
```suggestion
            .filter { it.hasSameBeforeEvaluationWith(id) }
            .forEach { it.resetBeforeEvaluation() }
```

로 수정해보았습니다.

지금처럼 그냥 stream 혹은 이전에 제이슨이 추천해주신 asSequence 그리고 기본적인 for-loop라는 세가지 선택지가 있었습니다.
정량적인 비교를 위해 measureNanoTime를 이용해보았습니다.
| 현상황 | asSequence | for-loop |
|----|----|----|
| 88472 | 6594927 | 75724  |

수정된 코드를 for-loop와의 차이가 별로 나지 않기에 보다 안전하다고 생각하는 스트림을 이용한 수정된 코드를 유지하도록 하겠습니다.

[참고글입니다](https://hwan-shell.tistory.com/245)